### PR TITLE
[hwloc] new recipe

### PR DESCRIPTION
**Pull request by ilya-lavrenov**
Specify library name and version:  **hwloc/2.9.0**

Closes https://github.com/conan-io/conan-center-index/issues/16973

Required for onetbb to support NUMA architectures / Hybrid cores.

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
@SSE4 @SpaceIm
could you please review and merge?

**Review by SSE4 — APPROVED** [review]

File: recipes/hwloc/all/conanfile.py
Comment (Nekto89) [inline recipes/hwloc/all/conanfile.py:15]: I think that it is closer to BSD-3-Clause

https://www.open-mpi.org/projects/hwloc/license.php
https://spdx.org/licenses/BSD-3-Clause.html

```suggestion
    license = "BSD-3-Clause"
```

Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: thanks, fixed!

File: recipes/hwloc/all/conanfile.py
Comment (Nekto89) [inline recipes/hwloc/all/conanfile.py:108]: post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library './lib/libhwloc.dylib' links to system library 'Foundation' but it is not in cpp_info.frameworks.

File: recipes/hwloc/all/conanfile.py
Comment (Nekto89) [inline recipes/hwloc/all/conanfile.py:104]: are all these properties needed?

File: recipes/hwloc/all/conanfile.py
Comment (Nekto89) [inline recipes/hwloc/all/conanfile.py:112]: isn't this automatically propagated from self.requires call if we don't use components?

**Review by Nekto89 — CHANGES_REQUESTED** [review]

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py:65]: good catch, thanks!

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: removed, since they should be auto-filled.

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: good point, removed.

File: recipes/hwloc/all/conanfile.py
Comment (Nekto89) [reply recipes/hwloc/all/conanfile.py:106]: "dylib" is created with self.options.shared==True

**Comment by ilya-lavrenov** [comment]
Can we merge the PR?

File: recipes/hwloc/all/conanfile.py
Comment (memsharded) [inline recipes/hwloc/all/conanfile.py:24]: Options with default ``False`` tend to behave more intuitively than the opposite.
Given that the build system provides ``HWLOC_WITH_LIBXML2``, I would provide a ``with_libxml2`` option, with a default=``False``

File: recipes/hwloc/all/conanfile.py
Comment (memsharded) [inline recipes/hwloc/all/conanfile.py:68]: AutotoolsDeps is not the best option for autotools dependencies, but ``PkgConfigDeps`` should be preferred in the general case. As the default doesn't include dependencies, it is difficult to know if this is working or not, did you test it locally? Please try to change it if possible.

File: recipes/hwloc/all/test_v1_package/CMakeLists.txt
Comment (memsharded) [inline recipes/hwloc/all/test_v1_package/CMakeLists.txt:1]: You can drop ``test_v1_package`` all together, it is no longer required for ConanCenter.

File: recipes/hwloc/all/conanfile.py
Comment (memsharded) [inline recipes/hwloc/all/conanfile.py:118]:
```suggestion
```

These can be dropped for new recipes. Only necessary for backward compat of previously existing recipes, but new ones should already use ``CMakeDeps`` to consume this.

**Review by memsharded — COMMENTED** [review]
This looks very solid, good job, thanks!
Please have a look to the comments, I think it is almost there.

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: Applied comments.
Yes, I have tested with `with_libxml2=True/False` on macOS, Linux.

**Comment by ilya-lavrenov** [comment]
@jcar87 @memsharded 
Could you please review and merge?

File: recipes/hwloc/all/conanfile.py
Comment (jcar87) [inline recipes/hwloc/all/conanfile.py:38]: if this is a pure C library, it should probably reflect so by removing the `cppstd` and `libcxx` compiler settings
https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py#L67C1-L68

File: recipes/hwloc/all/conanfile.py
Comment (jcar87) [inline recipes/hwloc/all/conanfile.py:73]: mm I think I can understand disabling cairo in the same vein as disabling opencl and cuda - but what does `--disable-io` does and why is it needed?

File: recipes/hwloc/all/conanfile.py
Comment (jcar87) [inline recipes/hwloc/all/conanfile.py:98]: in what context may these tools be used? what is the rationale behind removing them altogether?

File: recipes/hwloc/all/conanfile.py
Comment (jcar87) [inline recipes/hwloc/all/conanfile.py:102]: which files are typically stored in this location by this package, and why are they not needed?

File: recipes/hwloc/all/conanfile.py
Comment (jcar87) [reply recipes/hwloc/all/conanfile.py:108]: It actually doesn't have to if it's truly private - the warning is there as building the shared variant gives us a hint as to which libraries would be needed to be explicitly linked by consumers that use the static variant

File: recipes/hwloc/all/test_package/CMakeLists.txt
Comment (jcar87) [inline recipes/hwloc/all/test_package/CMakeLists.txt:1]: while not strictly necessary syntax-wise in this file, we are trying to make sure new recipes follow the template  - otherwise it's very likely that new PRs that touch this recipe will amend this, adding unnecessary review burden

please make sure this lists 3.15:
https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/test_package/CMakeLists.txt#L1

File: recipes/hwloc/all/test_package/test_package.c
Comment (jcar87) [inline recipes/hwloc/all/test_package/test_package.c:9]: I would suggest reducing the test to calling only `hwloc_topology_init` and `hwloc_topology_destroy`, as I can see those are functions that are inside the library.

Since this file is not checking for the int output of any of the functions, and it's explicitly returning 0 at the end of the control flow, it's superfluous to call any other functionality, since  we are not testing correct behaviour. This is good - where possible, we want the test_package to test the minimum necessary to make sure that we can compile and link. i believe that is achieved by the first function altogether and there's no need for more.

**Review by jcar87 — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution, this is a really nice and clean recipe.

I've added some comments with a request for some minor changes.

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py:73]:
```
  --disable-io            Disable I/O discovery build entirely (PCI, LinuxIO,
                          CUDA, OpenCL, NVML, RSMI, LevelZero, GL) instead of
                          only disabling it at runtime by default
```

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py:98]: These tools are kind of prebuilt samples, which you can use to play with hwloc functionality.
E.g. understand topology of your system, bind threads to cores, etc. I can hardly imagine that someone needs this during build of another application. 

If a developer plays with HWLOC, it can download archive and build on his / her own, while Conan recipe is useful for integration purposes, where these tools are not needed.

Even someone needs them in future, she / he can send a PR with changes where some tools are kept back and provide arguments why it's needed. For now, since I'm not hHWLOC developer / maintainer / advanced user, I'm not the right person for these questions. What I need is to enable topology knowledge for TBB, where minimal HWLOC capabilities are required.

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py:102]:
```
% tree -L 2 share
share
├── applications
│   └── lstopo.desktop
├── bash-completion
│   └── completions
├── doc
│   └── hwloc
├── hwloc
│   ├── hwloc-ps.www
│   ├── hwloc-valgrind.supp
│   ├── hwloc.dtd
│   ├── hwloc2-diff.dtd
│   └── hwloc2.dtd
└── man
    ├── man1
    ├── man3
    └── man7

12 directories, 5 files
```

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py:38]: added, good catch!

**Comment by ilya-lavrenov** [comment]
Can we finally merge?

File: recipes/hwloc/all/conanfile.py
Comment (RubenRBS) [inline recipes/hwloc/all/conanfile.py:15]: Topics that just repeat the library name are redundant (It's used for the search terms in conan center index), ones that use keywords that describe the library are generally much better :)

**Review by RubenRBS — COMMENTED** [review]
Thanks a lot for your contribution. I only have one minor comment about the topic

File: recipes/hwloc/all/conanfile.py
Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py:15]: do we have a guarantee that after fixing this comment the recipe will be merged?

I have already spent a week with fixing _minor_ comments from different reviewers (note, that recipe is green for a while). It's very annoying .. why cannot maintainers write all comments at once? Why every new person comes to the PR every two days and imagines more and more minor comments? Is it never ending loop?

Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: removed `topic` at all

Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: great.. this useless `topics` attribute is mandatory
Really? Who actually uses this tags? Everybody searches in google, not via Conan topics.

File: recipes/hwloc/all/conanfile.py
Comment (RubenRBS) [reply recipes/hwloc/all/conanfile.py:100]: Thanks a lot for the explanation, we appreciate it

File: recipes/hwloc/all/conanfile.py
Comment (RubenRBS) [reply recipes/hwloc/all/conanfile.py]: > why cannot maintainers write all comments at once? Why every new person comes to the PR every two days and imagines more and more minor comments? Is it never ending loop?

Let's discuss this better in your issue that deals about how you feel about the review process, I think it's a better place :)

> do we have a guarantee that after fixing this comment the recipe will be merged?

But no. If the remaining reviewer(s) find something else I might have missed, the review will continue, it's the whole point of the reviews! Again, let's better discuss this in the linked issue

> great.. this useless topics attribute is mandatory

I understand your frustration about what feel like minor stylistic choices that have nothing to do with the really important parts. We're actively working on improving contributor flows for such things and we appreciate the input, but yeah as I have mentioned, topics *are* used as a search term in Conan Center, and it's still a feature we want to support, that's why we ask for them to be useful :)

Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: > topics are used as a search term in Conan Center, and it's still a feature we want to support, that's why we ask for them to be useful in most cases :)

If you want this to be a Conan feature, anyway you will have to review all other recipes on topic attribute, because right now other recipes has the same duplication. 

> But no. If the remaining reviewer(s) find something else I might have missed, the review will continue

How many reviewers should review? Could _you_ please approve the PR? You are the 4th one who wrote comments, but nobody from the previous ones either finally approved the PR or wrote new comments.

Comment (RubenRBS) [reply recipes/hwloc/all/conanfile.py]: > because right now other recipes has the same duplication.

See the other issue where I touch upon this. Topics have changed over time, and old recipes that get not too many updates end up desynced with our current thinkings, and that's not a bad thing. Don't focus on old recipes having outdated criteria, it would hamper our ability to evolve in the end

> How many reviewers should review? 

> How is a final decision [made]?

See also the linked issue where I explain this a bit better, but you're right in that it should be made more clear what the expectations for reviews are, whose voice matters in the end, and how the process flows

> is it a sane strategy?

Please understand that behind every decision and review there are real people and we are sometimes fallible, and we might disagree on what the best way forward is


Having said that, I think we have discussed topics a lot. We'll internally review your comments about the process and will let you know once we act on those, but for now please add _some_ topics so that we can approve the PR and get it finally merged, thanks! :)

Comment (ilya-lavrenov) [reply recipes/hwloc/all/conanfile.py]: > some topics so that we can approve the PR and get it finally merged, thanks! :)

Already, hope this helps to get your 👍

Comment (RubenRBS) [reply recipes/hwloc/all/conanfile.py]: Thanks a lot!

**Comment by ilya-lavrenov** [comment]
@jcar87 @Nekto89 @memsharded 
You have provided review, but have not returned back to check how your comments were applied. Please, do it and don't block contributors.

**Comment by Nekto89** [comment]
> @jcar87 @Nekto89 @memsharded @RubenRBS You have provided review, but have not returned back to check how your comments were applied. Please, do it and don't block contributors.

My approval doesn't mean anything for merge process - I don't have "write access".

**Review by Nekto89 — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
> > @jcar87 @Nekto89 @memsharded @RubenRBS You have provided review, but have not returned back to check how your comments were applied. Please, do it and don't block contributors.
> 
> My approval doesn't mean anything for merge process - I don't have "write access".

But since you have provided meaningful comments, it's important to know that they are fixed and you don't have new ones :) 
Thanks!

**Review by jcar87 — APPROVED** [review]
Thanks for your contribution and for addressing the review comments.

I am really sorry about the delay in this getting merged. This is a really good recipe. However, I am hugely disappointed in the tone you are using to vent your frustrations.

**Review by RubenRBS — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
> Thanks for your contribution and for addressing the review comments.
> 
> I am really sorry about the delay in this getting merged. This is a really good recipe. However, I am hugely disappointed in the tone you are using to vent your frustrations.

Really sorry for that, nothing personal, but what do you expect when people are ignored and have to ask everyday to review their PRs? And this is happening with most of my (and as it turned out not only my) PRs. 

I think of it as:
- I spend many evenings learning Conan documentation (which is really not easy, especially taking into account that 1.x and 2.x versions have different documents)
- I wrote a good recipe for the library [requested](https://github.com/conan-io/conan-center-index/issues/16973) by your team 
- So, I have made all the steps as a contributor and expect Conan team will make its steps and provide review. But this is not happening and results of sleepless nights can be abandoned. Do you think I should happy with it? I'm disappointed more than you

**Pull request merged**
